employee-rostering-gwtui/src/main/java/org/optaweb/employeerostering/gwtui/client/pages/rotation/ShiftTemplateEditForm.java

```diff
@@ -0,0 +1,248 @@
+    }
+
+    @Override
+    protected void onClose() {
+        if (shiftTemplateGridObject != null) {
+            shiftTemplateGridObject.setSelected(false);
+        }
+    }
+
+    @EventHandler("apply-button")
+    public void onApplyButtonClick(@ForEvent("click") final MouseEvent e) {
+        if (shiftTemplateGridObject != null) {
+
+            final ShiftTemplateView template = shiftTemplateGridObject.getShiftTemplateModel().getShiftTemplateView();
+
+            if (updateShiftTemplateFromWidgetsIfValid(template)) {
+                RotationRestServiceBuilder.updateShiftTemplate(template.getTenantId(), template, callbackFactory.onSuccess((final ShiftTemplateView updatedShiftTemplate) -> {
+                    shiftTemplateGridObject.getShiftTemplateModel().withShiftTemplateView(updatedShiftTemplate);
+                    shiftTemplateGridObject.setSelected(false);
+                    eventManager.fireEvent(ROTATION_INVALIDATE);
+                    hide();
+                }));
+            }
+        } else {
+            throw new IllegalStateException("shiftTemplateGridObject is null.");
+        }
+
+        e.stopPropagation();
+    }
+
+    @EventHandler("delete-button")
+    public void onDeleteButtonClick(@ForEvent("click") final MouseEvent e) {
+        final ShiftTemplateView shiftTemplateView = shiftTemplateGridObject.getShiftTemplateModel().getShiftTemplateView();
+
+        RotationRestServiceBuilder.removeShiftTemplate(shiftTemplateView.getTenantId(), shiftTemplateView.getId(), callbackFactory.onSuccess(success -> {
+            if (success) {
+                shiftTemplateGridObject.getLane().removeGridObject(shiftTemplateGridObject.getShiftTemplateModel());
+                hide();
+            }
+        }));
+        e.stopPropagation();
+    }
+
+    public boolean updateShiftTemplateFromWidgetsIfValid(ShiftTemplateView template) {
+        final Long oldSpot = template.getSpotId();
+        final Long oldEmployee = template.getRotationEmployeeId();
+        final Duration oldDurationFromReference = template.getDurationBetweenRotationStartAndTemplateStart();
+        final Duration oldDurationOfTemplate = template.getShiftTemplateDuration();
+
+        try {
+            if (!from.reportValidity()) {
+                throw new ValidationException("Rotation shift's start day offset/time is invalid.");
+            }
+            if (!to.reportValidity()) {
+                throw new ValidationException("Rotation shift's end day offset/time is invalid.");
+            }
+            template.setSpotId(parseId(spotSelect.getSelectedValue()));
+            template.setRotationEmployeeId(parseId(employeeSelect.getSelectedValue()));
+            Duration newDurationFromReference = Duration.ofDays(from.getDayOffset()).plusSeconds(from.getTime().toSecondOfDay());
+            Duration newDurationOfTemplate;
+            if (to.getDayOffset() >= from.getDayOffset()) {
```

Comment:
@Christopher-Chianelli if you move the logic of computing the `newDurationOfTemplate` into a new method, you can easily unit test it. Currently, the else branch is not covered and contains interesting logic.

Reply:
Good idea; done